armv7a9-zynq7000/low.c

```diff
@@ -0,0 +1,439 @@
+
+#include "../low.h"
+#include "../plostd.h"
+#include "../serial.h"
+#include "../timer.h"
+#include "../syspage.h"
+#include "../phoenixd.h"
+
+
+struct{
+	u16 timeout;
+	u32 kernel_entry;
+	} low_common;
+
+
+/* Board command definitions */
+const cmd_t board_cmds[] = {
+	{ cmd_loadPL,    "fpga ", " - load binary bitstream - TBD" },
+	{ NULL, NULL, NULL }
+};
+
+
+const cmd_device_t devices[] = {
+	{ "com1", PDN_COM1 },
+	{ NULL, NULL }
+};
+
+
+/* Initialization functions */
+
+void low_init(void)
+{
+	_zynq_init();
+
+	interrupts_init();
+
+	syspage_init();
+
+	syspage_setAddress((void *)SYSPAGE_ADDRESS);
+
+	low_common.timeout = 3;
+	low_common.kernel_entry = 0;
+}
+
+
+void low_done(void)
+{
+
+}
+
+
+void low_initphfs(phfs_handler_t *handlers)
+{
+	handlers[PDN_COM1].open = phoenixd_open;
+	handlers[PDN_COM1].read = phoenixd_read;
+	handlers[PDN_COM1].write = phoenixd_write;
+	handlers[PDN_COM1].close = phoenixd_close;
+	handlers[PDN_COM1].dn = PHFS_SERIAL_LOADER_ID;
+}
+
+
+void low_initdevs(cmd_device_t **devs)
+{
+	*devs = (cmd_device_t *)devices;
+}
+
+
+void low_appendcmds(cmd_t *cmds)
+{
+	int i = 0;
+
+	/* Find the last declared cmd */
+	while (cmds[i++].cmd != NULL);
+
+	if ((MAX_COMMANDS_NB - --i) < (sizeof(board_cmds) / sizeof(cmd_t)))
+		return;
+
+	low_memcpy(&cmds[i], board_cmds, sizeof(board_cmds));
+}
+
+
+
+/* Setters and getters for common data */
+
+void low_setDefaultIMAP(char *imap)
+{
+	low_memcpy(imap, "ddr", 4);
+}
+
+
+void low_setDefaultDMAP(char *dmap)
+{
+	low_memcpy(dmap, "ddr", 4);
+}
+
+
+void low_setKernelEntry(u32 addr)
+{
+	u32 offs;
+
+	if ((u32)VADDR_KERNEL_INIT != (u32)ADDR_DDR) {
+		offs = addr - VADDR_KERNEL_INIT;
+		addr = ADDR_DDR + offs;
+	}
+
+	low_common.kernel_entry = addr;
+}
+
+
+void low_setLaunchTimeout(u32 timeout)
+{
+	low_common.timeout = timeout;
+}
+
+
+u32 low_getLaunchTimeout(void)
+{
+	return low_common.timeout;
+}
+
+
+addr_t low_vm2phym(addr_t addr)
+{
+	u32 offs;
+	if ((u32)VADDR_KERNEL_INIT != (u32)ADDR_DDR) {
+		offs = addr - VADDR_KERNEL_INIT;
+		addr = ADDR_DDR + offs;
+	}
+
+	return addr;
+}
+
+
+/* Functions modify registers */
+
+u8 low_inb(u16 addr)
+{
+	//TODO
+	return 0;
+}
+
+
+void low_outb(u16 addr, u8 b)
+{
+	//TODO
+}
+
+
+u32 low_ind(u16 addr)
+{
+	//TODO
+	return 0;
+}
+
+
+void low_outd(u16 addr, u32 d)
+{
+	//TODO
+}
+
+
+
+/* Functions make operations on memory */
+
+void low_setfar(u16 segm, u16 offs, u16 v)
+{
+	//TODO
+}
+
+
+u16 low_getfar(u16 segm, u16 offs)
+{
+	//TODO
+	return 0;
+}
+
+
+void low_setfarbabs(u32 addr, u8 v)
+{
+	//TODO
+}
+
+
+u8 low_getfarbabs(u32 addr)
+{
+	//TODO
+	return 0;
+}
+
+
+void low_setfarabs(u32 addr, u32 v)
+{
+	//TODO
+}
+
+
+u32 low_getfarabs(u32 addr)
+{
+	//TODO
+	return 0;
+}
+
+
+void low_copyto(u16 segm, u16 offs, void *src, unsigned int l)
+{
+	//TODO
+}
+
+
+void low_copyfrom(u16 segm, u16 offs, void *dst, unsigned int l)
+{
+	//TODO
+}
+
+
+void low_memcpy(void *dst, const void *src, unsigned int l)
```

Comment:
I suggest you take `low_memcpy()`  from  `imxrt117x`

Reply:
I have taken pattern from libphoenix's function which was written for armv7a:
https://github.com/phoenix-rtos/libphoenix/blob/1668ad951812697966a271a4cb7a50f63aa8c139/arch/armv7a/string.c#L22-L46

Have you found there some disadvantages?